p2p/transport/tcpreuse/listener_test.go

```diff
@@ -0,0 +1,465 @@
+			require.NoError(t, err)
+			wh := wsHandler{conns: make(chan *websocket.Conn, acceptQueueSize)}
+			go func() {
+				http.Serve(manet.NetListener(l), wh)
+			}()
+			go func() {
+				d := websocket.Dialer{}
+				for i := 0; i < N; i++ {
+					go func() {
+						ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+						defer cancel()
+						conn, _, err := d.DialContext(ctx, fmt.Sprintf("ws://%s", l.Addr().String()), http.Header{})
+						if err != nil {
+							t.Error("failed to dial", err, i)
+							return
+						}
+						err = conn.WriteMessage(websocket.TextMessage, []byte("hello"))
+						if err != nil {
+							t.Error(err)
+						}
+						_, _, err = conn.ReadMessage()
+						if err == nil {
+							t.Error("expected EOF got nil")
+						}
+					}()
+				}
+			}()
+			var wg sync.WaitGroup
+			for i := 0; i < N; i++ {
+				c := <-wh.conns
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					defer c.Close()
+					msgType, buf, err := c.ReadMessage()
+					if !assert.NoError(t, err) {
+						return
+					}
+					if !assert.Equal(t, msgType, websocket.TextMessage) {
+						return
+					}
+					if !assert.Equal(t, "hello", string(buf)) {
+						return
+					}
+				}()
+			}
+			wg.Wait()
+		})
+
+		t.Run(fmt.Sprintf("WebSocketTLS-reuseport:%v", disableReuseport), func(t *testing.T) {
+			cm := NewConnMgr(disableReuseport, nil, nil)
+			l, err := cm.DemultiplexedListen(listenAddr, DemultiplexedConnType_TLS)
+			require.NoError(t, err)
+			defer l.Close()
+			wh := wsHandler{conns: make(chan *websocket.Conn, acceptQueueSize)}
+			go func() {
+				s := http.Server{Handler: wh, TLSConfig: selfSignedTLSConfig(t)}
+				s.ServeTLS(manet.NetListener(l), "", "")
+			}()
+			go func() {
+				d := websocket.Dialer{TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}
+				for i := 0; i < N; i++ {
+					go func() {
+						ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+						defer cancel()
+						conn, _, err := d.DialContext(ctx, fmt.Sprintf("wss://%s", l.Addr().String()), http.Header{})
+						if err != nil {
+							t.Error("failed to dial", err, i)
+							return
+						}
+						err = conn.WriteMessage(websocket.TextMessage, []byte("hello"))
+						if err != nil {
+							t.Error(err)
+						}
+						_, _, err = conn.ReadMessage()
+						if err == nil {
+							t.Error("expected EOF got nil")
+						}
+					}()
+				}
+			}()
+			var wg sync.WaitGroup
+			for i := 0; i < N; i++ {
+				c := <-wh.conns
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					defer c.Close()
+					msgType, buf, err := c.ReadMessage()
+					if !assert.NoError(t, err) {
+						return
+					}
+					if !assert.Equal(t, msgType, websocket.TextMessage) {
+						return
+					}
+					if !assert.Equal(t, "hello", string(buf)) {
+						return
+					}
+				}()
+			}
+			wg.Wait()
+		})
+	}
+}
+
+func TestListenerMultiplexed(t *testing.T) {
+	listenAddr := ma.StringCast("/ip4/0.0.0.0/tcp/0")
+	const N = 20
```

Comment:
For some reason this breaks with N > 60 on macOS. Just making more than 150 concurrent connections breaks with a raw tcp listener and dialer. So this is probably not a bug in the code somewhere.